jmdaemon/jmdaemon/onionmc.py

```diff
@@ -0,0 +1,1210 @@
+            log.debug("Invalid message type, ignoring: {}".format(msgtype))
+            return
+
+        # real JM message; should be: from_nick, to_nick, cmd, message
+        try:
+            nicks_msgs = msgval.split(COMMAND_PREFIX)
+            from_nick, to_nick = nicks_msgs[:2]
+            msg = COMMAND_PREFIX + COMMAND_PREFIX.join(nicks_msgs[2:])
+            if to_nick == "PUBLIC":
+                #log.debug("A pubmsg is being processed by {} from {}; it "
+                #    "is {}".format(self.self_as_peer.nick, from_nick, msg))
+                self.on_pubmsg(from_nick, msg)
+                if self.self_as_peer.directory:
+                    self.forward_pubmsg_to_peers(msg, from_nick)
+            elif to_nick != self.nick:
+                if not self.self_as_peer.directory:
+                    log.debug("Ignoring message, not for us: {}".format(msg))
+                else:
+                    self.forward_privmsg_to_peer(to_nick, msg, from_nick)
+            else:
+                self.on_privmsg(from_nick, msg)
+        except Exception as e:
+            log.debug("Invalid joinmarket message: {}, error was: {}".format(
+                msgval, repr(e)))
+            return
+
+    def forward_pubmsg_to_peers(self, msg: str, from_nick: str) -> None:
+        """ Used by directory nodes currently. Takes a received
+        message that was PUBLIC and broadcasts it to the non-directory
+        peers.
+        """
+        assert self.self_as_peer.directory
+        pubmsg = self.get_pubmsg(msg, source_nick=from_nick)
+        msgtype = JM_MESSAGE_TYPES["pubmsg"]
+        # NOTE!: Specifically with forwarding/broadcasting,
+        # we introduce the danger of infinite re-broadcast,
+        # if there is more than one party forwarding.
+        # For now we are having directory nodes not talk to
+        # each other (i.e. they are *required* to only configure
+        # themselves, not other dns). But this could happen by
+        # accident.
+        encoded_msg = OnionCustomMessage(pubmsg, msgtype)
+        for peer in self.get_connected_nondirectory_peers():
+            # don't loop back to the sender:
+            if peer.nick == from_nick:
+                continue
+            log.debug("Sending {}:{} to nondir peer {}".format(
+                msgtype, pubmsg, peer.peer_location()))
+            self._send(peer, encoded_msg)
+
+    def forward_privmsg_to_peer(self, nick: str, message: str,
+                                from_nick: str) -> None:
+        assert self.self_as_peer.directory
+        peer = self.get_peer_by_nick(nick)
+        if not peer:
+            log.debug("We were asked to send a message from {} to {}, "
+                      "but {} is not connected.".format(from_nick, nick, nick))
+            return
+        # The `message` passed in has format COMMAND_PREFIX||command||" "||msg
+        # we need to parse out cmd, message for sending.
+        _, cmdmsg = message.split(COMMAND_PREFIX)

```

Comment:
Ah, very good catch :) Unfortunately I made it slightly tougher than it might be because `get_privmsg` requires the command separated out, but it's not a big deal. Fixed in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1182/commits/f74b04352c06c6293976ae173964b34672ac1390